### PR TITLE
BFD-2436: Update Journey of a BFD Resource From Ingestion to Digestion Diagram

### PR DESCRIPTION
**Pull request by JeremyIR**
**JIRA Ticket:**
[BFD-2436](https://jira.cms.gov/browse/BFD-2436)

**User Story or Bug Summary:**

In an effort to have accurate, maintainable, and professional documentation across all BFD channels, the [confluence doc](https://confluence.cms.gov/display/BFD/The+Journey+of+a+BFD+Resource%3A+From+Ingestion+to+Digestion)  that depicts the ingestion and digestion of BFD data must be updated. 

The currently existing diagram is hard to read, partially inaccurate and does not communicate the information in the best way. 

This PR provides 3 new diagrams meant to replace the single diagram in the existing confluence page.

---

### What Does This PR Do?

In order to ensure the confluence document is accurate and presentable, the diagrams in PlantUML and their corresponding SVG output are available for review here.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

Do the diagrams adequately and concisely depict the ingestion and digestion of data for BFD in a way that it can be presented to an engineer or stakeholder?
* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [x] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-rda-ingestion.plantuml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+"rda-streaming-api" --> "rda-pipeline-source"  
+"rda-db" --> "rda-pipeline-source"
+"rda-pipeline-source" -> "rda-db": Dead Letter Queue (DLQ) Processing
+"rda-db" --> "rda-pipeline-source"
+"rda-pipeline-source" -> "rda-streaming-api": Run 'DLQGrpcRdaSource' job if needed
+"rda-streaming-api" -> "rda-pipeline-source"
+"rda-pipeline-source" -> "rda-streaming-api": GetVersion: stream of current API version
+"rda-streaming-api" --> "rda-pipeline-source"
+"rda-pipeline-source" -> "rda-db": Get starting sequence number to process claims
+"rda-db" --> "rda-pipeline-source"
```
Comment (brianburton) [inline docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-rda-ingestion.plantuml:20]: RDA pipeline queries a table in the `rda` schema of the BFD dataase to obtain the starting sequence number value.  Is `rda-db` here meant to refer to the BFD `rda` schema?  Or to the RDA API's own internal database?

If the former then the arrow for writing claims to the database should be changed to reflect that.  If the latter this should be changed to point to `bfd-db` instead.

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-rda-ingestion.plantuml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+"rda-pipeline-source" -> "rda-streaming-api": Open stream, and receive messages for processing/to ingest claims \n via RdaFissClaimLoadJob and RdaMcsClaimLoadJob
+"rda-streaming-api" --> "rda-pipeline-source"
+"rda-pipeline-source" -> "rda-pipeline-sink": Send data to sink for storage
+"rda-pipeline-sink" -> "bfd-db": Store fiss and mcs claim data for BFD consumption
+"bfd-db" --> "rda-pipeline-sink"
+"rda-pipeline-sink" --> "rda-pipeline-source"
+"rda-pipeline-source" -> "rda-streaming-api": Stop processing and close stream if: \n no more claims to process or there is an error 
+"rda-streaming-api"--> "rda-pipeline-source"
+"rda-pipeline-source" -> "rda-db": Any messages that can't be processed written to DLQ table
```
Comment (brianburton) [inline docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-rda-ingestion.plantuml:29]: This actually happens during processing, not after the stream has closed.

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-rda-ingestion.plantuml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+participant "rda-streaming-api"
+database "rda-db"
+participant "rda-pipeline-sink"
+database "bfd-db"
+note over "rda-streaming-api": Runs every 5 min in prod \n 1 hr in test
+"rda-pipeline-source" -> "rda-streaming-api": Smoke test to determine connectivity
+"rda-pipeline-source" -> "rda-db": Smoke test to determine connectivity
+"rda-streaming-api" --> "rda-pipeline-source"  
+"rda-db" --> "rda-pipeline-source"
+"rda-pipeline-source" -> "rda-db": Dead Letter Queue (DLQ) Processing
```
Comment (brianburton) [inline docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-rda-ingestion.plantuml:13]: Do we want to expand on what this processing entails?  It is similar to the regular processing but consists of these steps:

- read failed claims from DLQ
- process each failed claim from RDA API individually
- update database for each claim and mark as resolved or obsolete if possible

Comment (mjburling) [reply docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-rda-ingestion.plantuml:13]: Good question! The original intent may have been closer to something of a high level overview of the ingestion process, but the diagram already seems to be a little _too_ technical to remain at a high level. 

Let's try to get something closer to complete here.

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+s3 -> "bfd-pipeline": ETL/RIF Loader process triggered
+group ETL
+"bfd-pipeline" -> "bfd-pipeline": datasetMonitorListener interface instantiated \n in CcwRifLoadJob class to manage ETL
+"bfd-pipeline" -> "bfd-pipeline": ETL process parses manifest.xml file \n for specific resource data \n and checks for DB operation \n denoted  in manifest i.e. INSERT/UPDATE
+"bfd-pipeline" -> "bfd-db": DB operations applied for resource \n based on column and rif layout fhir mappings
+"bfd-db" --> "bfd-pipeline"
+"bfd-pipeline" -> "bfd-pipeline": Process marked as 'complete'
+end
+"bfd-pipeline" -> s3: Load status sent to S3
+s3 -> slack: Status sent \n to #BFD-notices channel
```
Comment (mjburling) [inline docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml:21]: Perhaps this is aspirational and how it _should_ work, but today...
* the bfd-pipeline service produces logs which are scraped and forwarded by splunk
* a scheduled splunk query runs every 15 minutes to find specific discovery/load messages for the previous 15 minutes
* splunk formats messages based on those indications of discovery/load to #bfd-notices

Make it your own, but, I think we can afford to be a _little_ lossier here and add some of the following:

* `participant splunk` to represent a sort of _composite_ participant representing e.g. splunk's various components of forwarders/indexers/search heads/ui/configuration
* `"bfd-pipeline" --> splunk: forwarded service logs`
* `splunk --> slack: discovery`
* `splunk --> slack: load`

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-rda-ingestion.plantuml (hunk 1 above)
Comment (JeremyIR) [reply docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-rda-ingestion.plantuml:20]: Ah yes, that's what I meant.

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-rda-ingestion.plantuml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+"rda-pipeline-source" -> "rda-streaming-api": Run 'DLQGrpcRdaSource' job if needed
+"rda-streaming-api" -> "rda-pipeline-source"
+"rda-pipeline-source" -> "rda-streaming-api": GetVersion: stream of current API version
+"rda-streaming-api" --> "rda-pipeline-source"
+"rda-pipeline-source" -> "rda-db": Get starting sequence number to process claims
+"rda-db" --> "rda-pipeline-source"
+"rda-pipeline-source" -> "rda-streaming-api": Open stream, and receive messages for processing/to ingest claims \n via RdaFissClaimLoadJob and RdaMcsClaimLoadJob
+"rda-streaming-api" --> "rda-pipeline-source"
+"rda-pipeline-source" -> "rda-pipeline-sink": Send data to sink for storage
+"rda-pipeline-sink" -> "bfd-db": Store fiss and mcs claim data for BFD consumption
```
[the remark this answers is not in the document]
Comment (JeremyIR) [reply docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-rda-ingestion.plantuml:24]: Good catch

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-rda-ingestion.plantuml (hunk 3 above)
Comment (JeremyIR) [reply docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-rda-ingestion.plantuml:13]: It's doable to add it in, and looks better, thx

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml (hunk 4 above)
Comment (JeremyIR) [reply docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml]: Added

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+participant "bfd-pipeline"
+participant splunk
+note over splunk: Specific Splunk Components: \n forwarders/indexers/search heads/ui/configuration
+participant slack
+database "bfd-db"
+ccw -> s3: Timestamped folder with: \n RIF files for a resource, manifest.xml \n placed in 'bfd-prod-etl' S3 bucket  
+s3 -> s3: Timestamped folder moved to \n 'Incoming' folder of bucket
+s3 -> splunk: New data discovered by splunk by quereying BFD-pipeline logs every 15 min
+splunk -> slack: Data discover status sent \n to #BFD-notices channel
+slack --> s3
```
Comment (lsmitchell) [inline docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml:14]: What's this step represent? In the diagram, it kinda looks like we're waiting for slack to respond before we continue, but the slack hook is basically a unidirectional, async notification that doesnt block our loading, right? Maybe its a limitation (or my misreading) of the flowchart, but its a tiny bit confusing.

Unless of course we actually do wait for some signal from the slack hook to start processing, in which case its fine!

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+s3 -> "bfd-pipeline": ETL/RIF Loader process triggered
```
Comment (lsmitchell) [inline docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml:15]: Are we intentionally glossing over the whole resource setup for the pipeline? (i.e. an instance is created, setup in the proper environment, etc) Is that encompassed by "process" vaguely here?

**Review by lsmitchell — COMMENTED** [review]
Our ccw loading diagram misses some details for synthetic data loading (like that we only load data that is <= today's date, and that we also read from the Synthetic incoming folder, but I can see us also just not mentioning syntehtic data in a high-level document.

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+"bfd-pipeline" -> "bfd-db": DB operations applied for resource \n based on column and rif layout fhir mappings
+"bfd-db" --> "bfd-pipeline"
+"bfd-pipeline" -> "bfd-pipeline": Process marked as 'complete' and load status is logged
+end
+"bfd-pipeline" -> splunk: Data load messages in pipeline logs made available for querying by splunk
+splunk -> "bfd-pipeline": Splunk discovers new data load from pipeline logs
+"bfd-pipeline" --> splunk
+splunk -> slack: Load status sent \n to #BFD-notices channel
+slack --> s3
+s3 -> s3: Timestamped folder for resource moved \n from 'Incoming' to the 'Done' folder
```
[the remark this answers is not in the document]
Comment (JeremyIR) [reply docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml:28]: Yes this is just meant to be a prod level diagram

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml (hunk 6 above)
Comment (JeremyIR) [reply docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml:14]: Yes its a unidirectional and async but there are multiple actors and so some steps have to be in slightly awkward chronological order. Like in the RDA version.

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+s3 --> "bfd-pipeline": ETL process continues if more \n resources still in 'Incoming' folder
+"bfd-pipeline" --> s3
```
[the remark this answers is not in the document]
Comment (JeremyIR) [reply docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml]: Added

File: docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml (hunk 8 above)
[the remark this answers is not in the document]
Comment (lsmitchell) [reply docs/assets/fhir-resource-ingestion-digestion-diagrams/bfd-pipeline-ccw-ingestion.plantuml:28]: We still load synthetic data in prod, but I get what you're saying.

**Review by lsmitchell — APPROVED** [review]

**Review by lsmitchell — APPROVED** [review]

**Review by brianburton — APPROVED** [review]

**Pull request merged**